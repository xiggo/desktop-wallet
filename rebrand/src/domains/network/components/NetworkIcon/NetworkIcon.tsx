import { Networks } from "@payvo/sdk";
import cn from "classnames";
import React from "react";

import { Circle, CircleProperties } from "@/app/components/Circle";
import { Icon } from "@/app/components/Icon";
import { Tooltip } from "@/app/components/Tooltip";
import { Size } from "@/types";

interface NetworkIconProperties {
	network?: Networks.Network;
	as?: React.ElementType;
	size: Size;
	className?: string;
	shadowClassName?: string;
	iconSize?: Size;
	showTooltip?: boolean;
	tooltipDarkTheme?: boolean;
	noShadow?: boolean;
	isCompact?: boolean;
}

const Placeholder = (properties: CircleProperties) => (
	<Circle
		data-testid="NetworkIcon__placeholder"
		className="border-theme-secondary-200 text-theme-secondary-500 dark:border-theme-secondary-700"
		{...properties}
	/>
);

export const NetworkIcon: React.VFC<NetworkIconProperties> = ({
	network,
	iconSize = "lg",
	className,
	showTooltip = true,
	tooltipDarkTheme,
	isCompact = false,
	...properties
}) => {
	if (!network) {
		return <Placeholder className={className} {...properties} />;
	}

	const getClassName = () => {
		if (className) {
			return className;
		}

		if (network.isLive()) {
			return "text-theme-primary-600 border-theme-primary-100 dark:border-theme-primary-600";
		}

		return "text-theme-secondary-700 border-theme-secondary-300 dark:border-theme-secondary-700";
	};

	const renderIcon = () => {
		const tickerIcon = <Icon data-testid="NetworkIcon__icon" name={network.ticker()} size={iconSize} />;

		if (isCompact) {
			return (
				<div
					aria-label={network.displayName()}
					data-testid={`NetworkIcon-${network.coin()}-${network.id()}`}
					className={cn("w-5 h-5 inline-flex items-center", getClassName())}
				>
					{tickerIcon}
				</div>
			);
		}

		return (
			<Circle
				aria-label={network.displayName()}
				data-testid={`NetworkIcon-${network.coin()}-${network.id()}`}
				className={getClassName()}
				{...properties}
			>
				{tickerIcon}
			</Circle>
		);
	};

	return (
		<Tooltip content={network.displayName()} disabled={!showTooltip} theme={tooltipDarkTheme ? "dark" : undefined}>
			{renderIcon()}
		</Tooltip>
	);
};
