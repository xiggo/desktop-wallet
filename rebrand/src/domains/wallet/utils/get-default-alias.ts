import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let counter = sameCoinWallets.length;

	if (counter === 0) {
		counter = 1;
	}

	while (profile.wallets().findByAlias(makeAlias(counter))) {
		counter++;
	}

	return makeAlias(counter);
};
