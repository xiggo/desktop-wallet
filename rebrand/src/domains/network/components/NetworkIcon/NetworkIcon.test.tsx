import { Networks } from "@payvo/sdk";
import userEvent from "@testing-library/user-event";
import React from "react";

import { NetworkIcon } from "./NetworkIcon";
import { availableNetworksMock } from "@/tests/mocks/networks";
import { render, screen } from "@/utils/testing-library";

let network: Networks.Network;

describe("NetworkIcon", () => {
	beforeEach(() => {
		network = availableNetworksMock[0];
	});

	it.each([true, false])("should render when isCompact = %s", (isCompact: boolean) => {
		const { asFragment } = render(<NetworkIcon network={network} size="lg" isCompact={isCompact} />, {});

		expect(screen.getByTestId(`NetworkIcon-${network.coin()}-${network.id()}`)).toHaveAttribute(
			"aria-label",
			network.displayName(),
		);
		expect(screen.getByTestId("NetworkIcon__icon")).toBeInTheDocument();

		expect(asFragment()).toMatchSnapshot();
	});

	it("should render with tooltip in the dark mode", () => {
		render(<NetworkIcon network={network} size="lg" tooltipDarkTheme />, {});

		userEvent.hover(screen.getByTestId(`NetworkIcon-${network.coin()}-${network.id()}`));

		expect(screen.getByRole("tooltip")).toHaveAttribute("data-theme", "dark");
	});

	it("should render with test network", () => {
		network = availableNetworksMock[1];

		render(<NetworkIcon size="lg" network={network} />, {});

		expect(screen.getByTestId(`NetworkIcon-${network.coin()}-${network.id()}`)).toHaveAttribute(
			"aria-label",
			network.displayName(),
		);
		expect(screen.getByTestId("NetworkIcon__icon")).toBeInTheDocument();
	});

	it("should render network with custom classname", () => {
		render(<NetworkIcon size="lg" network={network} className="test" />, {});

		expect(screen.getByTestId(`NetworkIcon-${network.coin()}-${network.id()}`)).toHaveAttribute(
			"aria-label",
			network.displayName(),
		);
		expect(screen.getByTestId("NetworkIcon__icon")).toBeInTheDocument();
	});
});
