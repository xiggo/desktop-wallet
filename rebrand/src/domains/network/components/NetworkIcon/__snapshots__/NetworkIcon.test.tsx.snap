// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkIcon should render when isCompact = false 1`] = `
<DocumentFragment>
  <div
    aria-label="ARK"
    class="Circle__CircleWrapper-sc-d5u41q-0 egSsGQ text-theme-primary-600 border-theme-primary-100 dark:border-theme-primary-600 ring-theme-background"
    data-testid="NetworkIcon-ARK-ark.mainnet"
  >
    <div
      class="sc-bdvvtL qGXhc"
      data-testid="NetworkIcon__icon"
      height="20"
      width="20"
    >
      <svg>
        ark.svg
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`NetworkIcon should render when isCompact = true 1`] = `
<DocumentFragment>
  <div
    aria-label="ARK"
    class="w-5 h-5 inline-flex items-center text-theme-primary-600 border-theme-primary-100 dark:border-theme-primary-600"
    data-testid="NetworkIcon-ARK-ark.mainnet"
  >
    <div
      class="sc-bdvvtL qGXhc"
      data-testid="NetworkIcon__icon"
      height="20"
      width="20"
    >
      <svg>
        ark.svg
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
